### ISSUE TITLE
The backward call hangs on Torch for CPU

### ISSUE THREAD

**Issue by ljaljushkin** [pytorch/pytorch#91547]
### 🐛 Describe the bug

Very tiny model with two convolutions hangs on loss.backward() call.

```python
UNet(
  (down_path): ModuleList(
    (0): UNetConvBlock(
      (block): Conv2d(3, 2, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    )
  )
  (last): Conv2d(2, 13, kernel_size=(1, 1), stride=(1, 1))
)
```
Snippet from stacktrace in GDB:
```
#0  0x00007fca0afed99f in __GI___poll (fds=0x7fc9ebbd0040, nfds=1, timeout=5000) at ../sysdeps/unix/sysv/linux/poll.c:29
```
It’s reproduced on torch with CPU only, GPU one works fine.

The batch size is also important. No hang with batch size = 1, but freezes with batch size = 4.
Profiling of the script for different batches shows different implementation called:
No hang with `_slow_conv2d_forward` and `aten::_slow_conv2d_backward`.
Hang happens with `aten::mkldnn_convolution` and `aten::convolution_backward`.

Snapshot of reproducing script:
```python
def main():
    train_set = MockDataset()
    train_loader = DataLoader(train_set, batch_size=4, num_workers=1, drop_last=True)
    model = UNet(n_classes=13)
    print(model)
    device = 'cpu'
    model.to(device)
    optimizer = SGD(model.parameters(), lr=1e-3)
    for epoch in range(1):
        model.train()
        for step, batch_data in enumerate(train_loader):
            inputs = batch_data[0].to(device)
            labels = batch_data[1].to(device)
            optimizer.zero_grad()
            outputs = model(inputs)
            print(outputs.shape, labels.shape)
            loss = nn.CrossEntropyLoss()(outputs, labels)
            print(f"\nBefore HANG {loss}\n")
            loss.backward()
            print("\nAFTER HANG\n")
```
CPU: Intel(R) Core™ i9-10980XE CPU @ 3.00GHz (can’t reproduce on Intel(R) Core™ i9-10920X CPU @ 3.50GHz)

The issue is reproduced in the docker when the python script is launched in this way:
```bash
sh -c "(script.sh > log.txt 2>&1; wait)" >&- 2>&- &
```

Content of the script.sh:
```bash
#!/bin/bash -xe
. venv/bin/activate
python3 main.py
```

[The full reproducer, GDB stack trace and profile log for different batches](https://gist.github.com/ljaljushkin/c6daf65ea0c6e119b9336c5edba09a0c)

Original issue from the forum: https://discuss.pytorch.org/t/backward-hangs-on-torch-for-cpu/169282

### Versions

Collecting environment information...
PyTorch version: 1.13.1+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.29
Is CUDA available: False
CUDA runtime version: 11.1.105
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 470.161.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.1+cpu <-------- also reproduced with torch==1.12.1+cpu and torch==1.9.1+cpu
[pip3] torchvision==0.14.1+cpu <-------- also reproduced with torchvision==0.13.1+cpu and torch==0.10.1+cpu
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @VitalyFedyunin @ejguan

**Comment by samdow** [comment]
Adding high pri for deadlock/hang

**Comment by ezyang** [comment]
this is probably the use of OMP before fork problem; try switching your multiprocessing to spawn

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/issues/3619#issuecomment-343372298

**Comment by ljaljushkin** [comment]
@ezyang, thank you for pointing out the relevant thread, some suggestions were helpful.
There are 2 workarounds to avoid the hang:

1. `OMP_NUM_THREADS=1 MKL_NUM_THREADS=1`
2. `torch.set_num_threads(1)`

`MKL_THREADING_LAYER=GNU` and `torch.multiprocessing.set_start_method('spawn')` didn't help.

Is it happening because gcc OpenMP is not fork-safe? 
Can I install torch from pypi with another OpenMP or the only way to build torch from sources?

**Comment by ezyang** [comment]
Anything using threads is not fork safe. It's surprising set start method spawn didn't help; I wonder if you ran it early enough in the program. I would not expect swapping OpenMP implementation to help.

**Comment by ljaljushkin** [comment]
I did it in that way. Please let me know if I should do it differently.
```python
if __name__ == '__main__':
    torch.multiprocessing.set_start_method('spawn')
    main()
```

**Comment by ezyang** [comment]
that looks like it should work. Did forkserver also not work

**Comment by ezyang** [comment]
also, if you have any code running outside of main that would cause problems too

**Comment by ljaljushkin** [comment]
> ```python
> torch.multiprocessing.set_start_method('spawn')
> ```

the same with `torch.multiprocessing.set_start_method('fork')` - still hangs

**Comment by mingfeima** [comment]
@ljaljushkin can this be reproduced on latest torch?

Also it would be helpful to run the code with `export MKLDNN_VERBOSE=2` to get additional log if this issue is related to oneDNN.

**Comment by ljaljushkin** [comment]
@mingfeima, collected logs with `MKLDNN_VERBOSE=2`.
It's reproduced on the latest stable torch==1.13.1+cpu ([log](https://gist.github.com/ljaljushkin/c6daf65ea0c6e119b9336c5edba09a0c#file-mkldnn_verbose_torch_1-13-1-cpu)) and on the latest nightly torch==2.0.0.dev20230112+cpu ([log](https://gist.github.com/ljaljushkin/c6daf65ea0c6e119b9336c5edba09a0c#file-mkldnn_verbose_torch_2-0-0-dev20230112-cpu))

**Comment by mingfeima** [comment]
The problem may be `OMP_THREAD_LIMIT`. if the omp thread loop size is larger the actually used threads, it will hang in barrier. Usually we would recommend set `OMP_NUM_THREADS` to the number of physical CPU cores. PyTorch will decide whether to go parallel and the grain_size for each operator internally.

@jingxu10 , could you please double check on our side whether this issue is introduced by `OMP_THREAD_LIMIT`. IF so, please also update the BKM from intel website.

**Comment by jingxu10** [comment]
please make sure `OMP_THREAD_LIMIT` is set to a number equal to or larger than `OMP_NUM_THREADS` to avoid the hanging issue.
